### ISSUE TITLE
Error while uploading AIML in the admin panel.

### ISSUE THREAD

**Issue by alihossain26** [Program-O/Program-O#142]
Hello, I have installed Program O version : 2.5.4. I have installed perfectly following the installation procedure and have successfully logged into the admin panel where I can upload or download AIML and much more things. But the problem is, When I goes to upload AIML (Currently ALICE's AIML), an error occurs saying "There was an error moving the file." and the files don't upload. i tried uploading using both zip file and single AIML file but doesn'
![problem](https://cloud.githubusercontent.com/assets/20319500/16617711/13774bd4-43a7-11e6-8355-208eca73086e.png)
t work. Please I need help. Thank you.

**Comment by Dave-Morton** [comment]
There are several reasons why that error might occur, so I'm going to need a bit more information before we can find the cause. The first place to look is in the admin page's error log file, located in the logs folder. Please take a look to see if there's anything in the file (logs/admin.error.log) that relates to this issue. Also, if you're having problems uploading a specific file, you can attach it here, and I can test the file itself to see if it's something that can be fixed within the file itself.

**Comment by alihossain26** [comment]
Thanks for your reply. I think there are many issues in the admin.error.log so I am uploading the file. And there is no specific file that is causing the problem but every AIML. It would be nice if you send me a AIML file so that I can test it. And I hope you would find the problem in the error log. Thank you.
[admin.error.docx](https://github.com/Program-O/Program-O/files/349919/admin.error.docx)

**Comment by Dave-Morton** [comment]
This AIML file has been verified as both well-fformed XML and valid AIML, and can be uploaded without a problem on my various different development installs of Program O:

[admin_added.aiml.txt](https://github.com/Program-O/Program-O/files/349936/admin_added.aiml.txt)

Try uploading that (after removing the .txt file extension, of course) while I look over your error logs to see what I can see.

**Comment by alihossain26** [comment]
Okay. Thank you very much. :+1:

**Comment by Dave-Morton** [comment]
It seems that you have a permissions issue on your server. I'm seeing several entries like this:

>[06-Jul-2016 12:20:21 UTC] PHP Warning:  mkdir(): Permission denied in /home/{username masked}/public_html/My AI/Program-O-master/admin/upload.php on line 281

This could be caused by your admin folder (or the files within it, such as upload.php) not having the correct permissions set. You might be able to fix the problem yourself through your FTP client, or you may have to ask your hosting provider for help with this. The root cause of this is that the script upload.php is being denied permission to create the uploads folder that will contain the AIML files that will be processed into the database. If you need specific assistance with changing permissions, please let me know and I'll assist you with the process once I get back from my day job. For now, I'm late for work.

**Comment by alihossain26** [comment]
Okay. I tried uploading the file but same problem "There was an error moving the file.".

**Comment by alihossain26** [comment]
Thank you very much for your help. I think I will need your help to get through the problem but it's no hurry. Do your work first. See you later then. Thanks again Dave.
Best regards,
Ali

**Comment by alihossain26** [comment]
Hello Dave, I have solved the problem changing the permission of "Config" folder. And the files are being uploaded successfully. Thank you very much for your help.

**Comment by alihossain26** [comment]
I will close the issue then.

**Comment by Jloupf** [comment]
Hi Ali,

I've had the same problem but i've found a solution.

Create a directory 'uploads' in your config directory. It worked for me.

**Comment by alihossain26** [comment]
Thanks Jloupf. I solved my problem. Thanks for commenting. :)

**Comment by jigs1212** [comment]
I have the same issue uploading the aiml files.

[27-Dec-2016 22:03:26 Asia/Calcutta] PHP Warning:  move_uploaded_file(./uploads/astrology.aiml): failed to open stream: Permission denied in /var/www/html/jigs-work/Program-O/admin/upload.php on line 336
[27-Dec-2016 22:03:26 Asia/Calcutta] PHP Warning:  move_uploaded_file(): Unable to move '/tmp/phpvC1e5V' to './uploads/astrology.aiml' in /var/www/html/-----/Program-O/admin/upload.php on line 336

**Comment by Dave-Morton** [comment]
@jigs1212 Other than the obvious fact that you're on a Linux server, we have very little to go on for helping you to figure out your issue. Please follow the instructions [here](https://github.com/Program-O/Program-O/wiki/Reporting-Issues) for providing information that helps us to help you.

Also, have you read _all_ of the posts in this issue? If so, and nothing in them has helped, then you're in the midst of a very strange problem, and we'll need a __LOT__ more information, including a detailed list of things you've tried, in addition to the other items that we need. If you haven't read all of the posts in this issue, then I strongly urge you to do so before you continue.

**Comment by jigs1212** [comment]
Sorry that i missed the System information.

Installed locally OS Lubuntu 16.04

Apache2 

mysql  Ver 14.14 Distrib 5.7.16, for Linux (x86_64) using  EditLine wrapper

PHP 7.0.8-0ubuntu0.16.04.3 (cli) ( NTS )

Chrome Version 55.0.2883.87 (64-bit)


[27-Dec-2016 22:03:26 Asia/Calcutta] PHP Warning: move_uploaded_file(./uploads/astrology.aiml): failed to open stream: Permission denied in /var/www/html/jigs-work/Program-O/admin/upload.php on line 336
[27-Dec-2016 22:03:26 Asia/Calcutta] PHP Warning: move_uploaded_file(): Unable to move '/tmp/phpvC1e5V' to './uploads/astrology.aiml' in /var/www/html/-----/Program-O/admin/upload.php on line 336

**Comment by jigs1212** [comment]
DEBUG FILE 


27-12-2016 21:40:48.93653[0][4] - Elapsed time: 1.199 milliseconds
[conversation_start.php][Called outside of function][49]
Conversation starting. Current system specs:
  Program O version:            2.6.3
  Server Software:              Apache/2.4.18 (Ubuntu)
  PHP Version:                  7.0.8-0ubuntu0.16.04.3
  OS:                           Linux
  OS Version:                   #78-Ubuntu SMP Fri Dec 9 23:50:32 UTC 2016
  MySQL Version:                5.7.16-0ubuntu0.16.04.1
  Multi-byte functions enabled: true 
-----------------------
27-12-2016 21:40:48.94743[4][4] - Elapsed time: 10.902 milliseconds
[load_aimlfunctions.php][Called outside of function][15]
AIML function include files loaded 
-----------------------
27-12-2016 21:40:48.95063[4][4] - Elapsed time: 3.203 milliseconds
[load_convofunctions.php][Called outside of function][15]
Convofunction include files loaded 
-----------------------
27-12-2016 21:40:48.95116[4][4] - Elapsed time: 0.528 milliseconds
[load_userfunctions.php][Called outside of function][13]
userfunctions include files loaded 
-----------------------
27-12-2016 21:40:48.95351[4][4] - Elapsed time: 2.353 milliseconds
[load_addons.php][Called outside of function][19]
Loading addons 
-----------------------
27-12-2016 21:40:48.95354[4][4] - Elapsed time: 0.029 milliseconds
[conversation_start.php][Called outside of function][58]
Loaded all Includes 
-----------------------
27-12-2016 21:40:48.95368[0][4] - Elapsed time: 0.139 milliseconds
[conversation_start.php][Called outside of function][100]
Debug level: 4
session ID = ablhtro2304dfanr6f0r4h5js4 
-----------------------
27-12-2016 21:40:48.9537[4][4] - Elapsed time: 0.023 milliseconds
[intialise_conversation.php][intialise_convoArray][28]
Intialising conversation 
-----------------------
27-12-2016 21:40:48.95372[4][4] - Elapsed time: 0.017 milliseconds
[intialise_conversation.php][load_blank_array_element][59]
Loading blank topic array 
-----------------------
27-12-2016 21:40:48.95374[4][4] - Elapsed time: 0.021 milliseconds
[intialise_conversation.php][load_blank_array_element][59]
Loading blank that array 
-----------------------
27-12-2016 21:40:48.95376[4][4] - Elapsed time: 0.017 milliseconds
[intialise_conversation.php][load_blank_array_element][59]
Loading blank star array 
-----------------------
27-12-2016 21:40:48.95378[4][4] - Elapsed time: 0.017 milliseconds
[intialise_conversation.php][load_blank_array_element][59]
Loading blank input array 
-----------------------
27-12-2016 21:40:48.95379[4][4] - Elapsed time: 0.017 milliseconds
[intialise_conversation.php][load_blank_stack][81]
Loading blank stack 
-----------------------
27-12-2016 21:40:48.95381[4][4] - Elapsed time: 0.016 milliseconds
[intialise_conversation.php][load_default_bot_values][105]
Loading db bot personality properties 
-----------------------
27-12-2016 21:40:48.95382[3][4] - Elapsed time: 0.016 milliseconds
[intialise_conversation.php][load_default_bot_values][109]
load db bot personality values SQL: SELECT * FROM `jigwis`.`botpersonality` WHERE `bot_id` = '1' 
-----------------------
27-12-2016 21:40:49.12311[2][4] - Elapsed time: 169.280 milliseconds
[handle_user.php][load_new_client_defaults][24]
Loading client defaults 
-----------------------
27-12-2016 21:40:49.12316[2][4] - Elapsed time: 0.051 milliseconds
[conversation_start.php][Called outside of function][218]
Conversation intialised waiting user 
-----------------------
27-12-2016 21:40:49.12331[2][4] - Elapsed time: 0.153 milliseconds
[conversation_start.php][Called outside of function][223]
Closing Database 
-----------------------
27-12-2016 21:40:49.12333[2][4] - Elapsed time: 0.016 milliseconds
[PDO_functions.php][db_close][49]
This DB is now closed. You don't have to go home, but you can't stay here. 
-----------------------
27-12-2016 21:40:49.12339[0][4] - Elapsed time: 0.062 milliseconds
[conversation_start.php][Called outside of function][228]
Conversation Ending. Elapsed time: 188.043 milliseconds. 
-----------------------
Debug Level: 4
-----------------------
Debug Mode: 1
-----------------------
CONVERSATION ARRAY
-----------------------
Array
(
    [conversation] => Array
        (
        )

    [topic] => Array
        (
            [1] => 
            [2] => 
            [3] => 
            [4] => 
            [5] => 
            [6] => 
            [7] => 
            [8] => 
            [9] => 
            [10] => 
            [11] => 
        )

    [that] => Array
        (
            [1] => 
            [2] => 
            [3] => 
            [4] => 
            [5] => 
            [6] => 
            [7] => 
            [8] => 
            [9] => 
            [10] => 
            [11] => 
        )

    [star] => Array
        (
            [1] => 
            [2] => 
            [3] => 
            [4] => 
            [5] => 
            [6] => 
            [7] => 
            [8] => 
            [9] => 
            [10] => 
            [11] => 
        )

    [input] => Array
        (
            [1] => 
            [2] => 
            [3] => 
            [4] => 
            [5] => 
            [6] => 
            [7] => 
            [8] => 
            [9] => 
            [10] => 
            [11] => 
        )

    [stack] => Array
        (
            [top] => om
            [second] => om
            [third] => om
            [fourth] => om
            [fifth] => om
            [sixth] => om
            [seventh] => om
            [last] => om
        )

    [bot_properties] => Array
        (
            [name] => Jigwis
            [gender] => Male
            [botmaster] => Jigs
            [master] => 
            [age] => 24
            [website] => jigwis.local
            [email] => mailme.jigs@gmail.com
            [birthday] => 12-08-2016
            [birthplace] => Calicut
            [size] => 
            [version] => 
            [build] => 
            [language] => English
            [feelings] => 
            [etype] => 
            [emotions] => 
            [ethics] => 
            [orientation] => 
            [baseballteam] => 
            [footballteam] => Germany
            [hockeyteam] => Indian
            [favoritesport] => 
            [vocabulary] => 
            [celebrities] => 
            [celebrity] => 
            [favoriteactor] => 
            [favoriteactress] => 
            [favoriteartist] => 
            [favoritemusician] => 
            [favoritesong] => 
            [favoriteauthor] => 
            [friend] => 
            [nationality] => 
            [religion] => 
            [president] => 
            [party] => 
            [kingdom] => 
            [phylum] => 
            [class] => 
            [order] => 
            [family] => 
            [genus] => 
            [species] => 
            [boyfriend] => 
            [favoritebook] => 
            [favoriteband] => 
            [favoritecolor] => 
            [favoritefood] => 
            [favoritemovie] => 
            [forfun] => 
            [friends] => 
            [girlfriend] => 
            [kindmusic] => 
            [location] => 
            [looklike] => 
            [question] => 
            [sign] => 
            [talkabout] => 
            [wear] => 
            [loves] => 
        )

    [client_properties] => Array
        (
            [name] => Seeker
            [id] => ablhtro2304dfanr6f0r4h5js4
        )

    [send_to_user] => User input not detected.
)

-----------------------
Total execution time: 188.043 Milliseconds. Goodbye.

**Comment by Dave-Morton** [comment]
No worries. While reviewing the posts in this issue, I ran across something that I must have previously missed.  @alihossain26 had mentioned that changing the permissions in the `config` folder had solved the issue, and that got me to thinking, so I did some digging, and noticed that in the file `admin/upload.php` the upload path is set inline to `./uploads/`, and this is incorrect. There's already a constant set in the config that describes the __absolute path__ for the uploads directory, and it wasn't being used that way, so I changed it in upload.php, and will be committing shortly, once I'm done testing. Meanwhile, you can make the change yourself if you like by opening `admin/upload.php` in your favorite (or least favorite, if you wish... I'm cool with that, too) text editor, find line 335, and change this:

    $file = './uploads/' . $_FILES['aimlfile']['name'];

To this:

    $file = _UPLOAD_PATH_ . $_FILES['aimlfile']['name'];

That should correct the issue.

**Comment by jigs1212** [comment]
Thanks for the quick reply but i am still facing the same issue.

**Comment by jigs1212** [comment]
#This time error log is different

[27-Dec-2016 23:28:14 Asia/Calcutta] PHP Warning:  move_uploaded_file(/var/www/html/jigs-work/Program-O/admin/uploads/client_profile.aiml): failed to open stream: Permission denied in /var/www/html/jigs-work/Program-O/admin/upload.php on line 339
[27-Dec-2016 23:28:14 Asia/Calcutta] PHP Warning:  move_uploaded_file(): Unable to move '/tmp/phpgPUMUA' to '/var/www/html/jigs-work/Program-O/admin/uploads/client_profile.aiml' in /var/www/html/jigs-work/Program-O/admin/upload.php on line 339

**Comment by Dave-Morton** [comment]
Have you verified the permissions for the uploads folder, by chance? It may not be a bad idea, also, to check the ownership of the folder, though it really shouldn't make a difference. Beyond that, there is no reason for this error to be occurring.

**Comment by jigs1212** [comment]
Thanks @Dave-Morton I have given 777 to uploads and all started working. Also updated the pathline in the upload folder.

**Comment by Dave-Morton** [comment]
For security purposes, 755 is all that's necessary. Execution of files in the uploads folder is a bad idea, really.

**Comment by jigs1212** [comment]
earlier it was 755 then I changed to 777 that's y it has worked now. then I guess I am going wrong somewhere with the permissions.

**Comment by Dave-Morton** [comment]
Interesting. If it fails with permissions of 755, that makes no sense. It would almost seem that the script was trying to execute a file from the uploads folder, but that should in no way happen. Let me look deeper into the code.

**Comment by Dave-Morton** [comment]
Ok, kindly ignore my last response. The problem seems to be that the _owner_ of the folder isn't your web server. I suggest checking that folder to see who "owns" it, and if it's not the same user/group of your web server (usually www-data by default), change the ownership via chown.

**Comment by jigs1212** [comment]
Roger that.

**Comment by Dave-Morton** [comment]
Any folder that gets written to (e.g. uploads, downloads, logs and debug) needs to have the same owner as your web server, so it stands to reason that if one needs to be changed, they _all_ should.

**Comment by jigs1212** [comment]
Ok thank you for the help :)

**Comment by Dave-Morton** [comment]
Since there's no further word on the status of this issue in more than 30 days, I have to assume that it's working as intended, so I'm closing this issue. If you're still experiencing trouble, please open a new issue. Thanks.
